MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java

```diff
@@ -212,43 +189,91 @@ static MotionPhotoReader open(File file,
      * set up for the video track to read frame data.
      */
-    @RequiresApi(api = 23)
-    private void startRenderThread(MotionPhotoInfo motionPhotoInfo) throws IOException {
+    private void startRenderThread(MotionPhotoInfo motionPhotoInfo, boolean stabilizationOn) throws IOException {
         // Set up the render handler and thread
         renderWorker = new HandlerThread("renderHandler");
         renderWorker.start();
         renderHandler = new Handler(renderWorker.getLooper());
+        bufferProcessor = new BufferProcessor(availableInputBuffers, availableOutputBuffers);
 
         // Set up input stream from Motion Photo file for media extractor
         fileInputStream = new FileInputStream(file);
         FileDescriptor fd = fileInputStream.getFD();
         int videoOffset = motionPhotoInfo.getVideoOffset();
-        lowResExtractor.setDataSource(fd, file.length() - videoOffset, videoOffset);
+        extractor.setDataSource(fd, file.length() - videoOffset, videoOffset);
+
+        // Find the do_not_stabilize bit in the image metadata track and set stabilizationOn
+        // 1. Check if the bit exists
+        //   a. If the bit exists, set stabilizationOn to true if it was originally true
+        //   b. If the bit does not exist, override stabilizationOn and set it to false
+        // 2. If the bit does not exist, then override stabilizationOn and set it to false
+        int version = motionPhotoInfo.getVersion();
+        if (version == MOTION_PHOTO_V1) {
+            for (int i = 0; i < extractor.getTrackCount(); i++) {
+                MediaFormat format = extractor.getTrackFormat(i);
+                String mime = format.getString(MediaFormat.KEY_MIME);
+                assert mime != null;
+                if (mime.startsWith(MOTION_PHOTO_IMAGE_META_MIMETYPE)) {
+                    Log.d(TAG, "selected image meta track: " + i);
+                    extractor.selectTrack(i);
+                    ByteBuffer inputBuffer = ByteBuffer.allocateDirect((int) extractor.getSampleSize());
+                    int sampleSize = extractor.readSampleData(inputBuffer, 0);
+                    if (sampleSize >= 0) {
+                        // The do_not_stabilize bit is available
+                        ImageMeta.ImageData imageData = ImageMeta.ImageData.parseFrom(inputBuffer);
+                        if (imageData.hasDoNotStabilize()) {
+                            stabilizationOn = stabilizationOn && !imageData.getDoNotStabilize();
+                        } else {
+                            stabilizationOn = false;
+                        }
+                    } else {
+                        // The do_not_stabilize bit is unavailable
+                        stabilizationOn = false;
+                    }
+                    extractor.unselectTrack(i);
+                    break;
+                }
+            }
+        }
```

Comment:
it would be easier to read if you can extract this into a separate private helper method. Prefer something like 
```
private boolean shouldStabilize(... ) { ... }
```

If it helps, use Android Studio's refactor to automatically determine what function inputs you need (i.e. select block of code, right click, refactor -> extract method) and then tweak the resulting extracted method.